File: 1388-바닥 장식/main.py
import sys
input = sys.stdin.readline

N, M = map(int, input().split())

board = []
visited = [[False] * M for _ in range(N)]
count = 0

for i in range(N):
  board.append(input())

def dfs(x: int, y: int, tile: str, upcount: bool = False):
  if x < 0 or y < 0 or x >= M or y >= N: return
  if visited[y][x]: return

  global count
  if upcount: count += 1

  if(board[y][x] == tile):
    visited[y][x] = True

    if tile == "-":
      dfs(x - 1, y, tile)
      dfs(x + 1, y, tile)
    elif tile == "|":
      dfs(x, y - 1, tile)
      dfs(x, y + 1, tile)

for i in range(N):
  for j in range(M):
    dfs(j, i, board[i][j], True)

print(count)

File: 1388-바닥 장식/main.cpp
#include <iostream>
#include <vector>
using namespace std;

int N, M;

void dfs(vector<vector<char>>& board, vector<vector<bool>>& visited, int& count,
         int x, int y, char tile, bool upcount) {
  if (x < 0 || y < 0 || x >= M || y >= N) return;
  if (visited[y][x]) return;

  if (upcount) count++;

  if (tile == board[y][x]) {
    visited[y][x] = true;

    if (tile == '-') {
      dfs(board, visited, count, x - 1, y, tile, false);
      dfs(board, visited, count, x + 1, y, tile, false);
    } else if (tile == '|') {
      dfs(board, visited, count, x, y - 1, tile, false);
      dfs(board, visited, count, x, y + 1, tile, false);
    }
  }
}

int main() {
  cin >> N >> M;

  vector<vector<char>> board(N, vector<char>(M));
  vector<vector<bool>> visited(N, vector<bool>(M, false));
  int count = 0;
  for (int i = 0; i < N; i++) {
    for (int j = 0; j < M; j++) {
      cin >> board[i][j];
    }
  }

  for (int i = 0; i < N; i++) {
    for (int j = 0; j < M; j++) {
      dfs(board, visited, count, j, i, board[i][j], true);
    }
  }

  cout << count;
}